File: production/frontend/src/api/feedback.ts
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? ''

export type FeedbackType = 'bug' | 'idea'
export type FeedbackStatus = 'new' | 'triaged' | 'done'

export interface FeedbackItem {
  id: number
  wedding_id: number
  submitted_by: string
  type: FeedbackType
  message: string
  page: string | null
  role: string | null
  viewport: string | null
  status: FeedbackStatus
  created_at: string
}

export interface FeedbackCreate {
  type: FeedbackType
  message: string
  page?: string | null
  role?: string | null
  viewport?: string | null
}

export class FeedbackApiError extends Error {
  readonly status: number | null

  constructor(message: string, status: number | null = null) {
    super(message)
    this.name = 'FeedbackApiError'
    this.status = status
  }
}

async function readErrorMessage(response: Response, fallback: string): Promise<string> {
  const payload = await response.json().catch(() => null)
  if (payload && typeof payload === 'object' && 'detail' in payload) {
    const detail = (payload as { detail: unknown }).detail
    if (typeof detail === 'string') {
      return detail
    }
  }

  return fallback
}

export async function createFeedback(
  payload: FeedbackCreate,
  apiBaseUrl = API_BASE_URL,
): Promise<FeedbackItem> {
  const response = await fetch(`${apiBaseUrl}/api/feedback`, {
    body: JSON.stringify(payload),
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    method: 'POST',
  })

  if (!response.ok) {
    throw new FeedbackApiError(
      await readErrorMessage(response, 'Unable to send your feedback.'),
      response.status,
    )
  }

  return response.json() as Promise<FeedbackItem>
}

export async function fetchFeedback(apiBaseUrl = API_BASE_URL): Promise<FeedbackItem[]> {
  const response = await fetch(`${apiBaseUrl}/api/feedback`, {
    credentials: 'include',
  })

  if (!response.ok) {
    throw new FeedbackApiError(
      await readErrorMessage(response, 'Unable to load feedback.'),
      response.status,
    )
  }

  return response.json() as Promise<FeedbackItem[]>
}

export async function updateFeedback(
  id: number,
  status: FeedbackStatus,
  apiBaseUrl = API_BASE_URL,
): Promise<FeedbackItem> {
  const response = await fetch(`${apiBaseUrl}/api/feedback/${id}`, {
    body: JSON.stringify({ status }),
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    method: 'PATCH',
  })

  if (!response.ok) {
    throw new FeedbackApiError(
      await readErrorMessage(response, 'Unable to update this feedback.'),
      response.status,
    )
  }

  return response.json() as Promise<FeedbackItem>
}

File: production/backend/app/api/feedback.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.auth import require_coordinator, require_guest
from app.api.schemas_auth import UserResponse
from app.db.database import get_db
from app.db.models import Feedback
from app.db.schemas import FeedbackCreate, FeedbackResponse, FeedbackUpdate
from app.logging import get_logger


router = APIRouter(prefix="/api/feedback", tags=["feedback"])
logger = get_logger(__name__)


def get_feedback_or_404(db: Session, feedback_id: int, wedding_id: int) -> Feedback:
    feedback = db.get(Feedback, feedback_id)
    if feedback is None or feedback.wedding_id != wedding_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Feedback not found"
        )
    return feedback


@router.post("", response_model=FeedbackResponse, status_code=status.HTTP_201_CREATED)
async def create_feedback(
    payload: FeedbackCreate,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(require_guest),
) -> Feedback:
    """Any signed-in user (guests included) can report a bug or share an idea.

    Deliberately not phase-gated: feedback should be welcome from the first
    invite preview to the last dance.
    """
    db_feedback = Feedback(
        wedding_id=current_user.wedding_id,
        submitted_by=current_user.name,
        type=payload.type,
        message=payload.message,
        page=payload.page,
        role=payload.role,
        viewport=payload.viewport,
    )
    db.add(db_feedback)
    db.commit()
    db.refresh(db_feedback)
    logger.info(
        "feedback_submitted",
        extra={"feedback_id": db_feedback.id, "feedback_type": db_feedback.type},
    )
    return db_feedback


@router.get("", response_model=list[FeedbackResponse])
async def list_feedback(
    status_filter: Literal["new", "triaged", "done"] | None = Query(
        None, alias="status"
    ),
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(require_coordinator),
) -> list[Feedback]:
    query = db.query(Feedback).filter(Feedback.wedding_id == current_user.wedding_id)
    if status_filter is not None:
        query = query.filter(Feedback.status == status_filter)
    return query.order_by(Feedback.created_at.desc(), Feedback.id.desc()).all()


@router.patch("/{feedback_id}", response_model=FeedbackResponse)
async def update_feedback(
    feedback_id: int,
    payload: FeedbackUpdate,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(require_coordinator),
) -> Feedback:
    feedback = get_feedback_or_404(db, feedback_id, current_user.wedding_id)
    feedback.status = payload.status
    db.commit()
    db.refresh(feedback)
    logger.info(
        "feedback_updated",
        extra={"feedback_id": feedback.id, "feedback_status": feedback.status},
    )
    return feedback
